mpi4: MPI_T Events

raffenet opened the pull request:
## Pull Request Description

Implementation of MPI_T Events.

<!-- AUTHOR: After creating this merge request, check off each of the following items as you complete them. -->

## Expected Impact

Add framework for exposing MPI library information as events.

## Author Checklist
* [x] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [x] Remove xfail from the test suite when fixing a test
* [x] Commits are self-contained and do not do two things at once
* [x] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [x] Passes whitespace checkers
* [x] Passes warning tests
* [x] Passes all tests
* [x] Add comments such that someone without knowledge of the code could understand
* [x] You or your company has a signed contributor's agreement on file with Argonne
* [x] For non-Argonne authors, request an explicit comment from your companies PR approval manager

raffenet commented:
This PR is, IMO, MPI-4.0 compliant in its current state. As an example event, I am thinking of expanding ch4 recvq PVAR counters to also support callbacks through the events interface. Hopefully can hack that up relatively quickly.

raffenet commented:
test:mpich/ch4/most

raffenet commented:
test:mpich/ch4/most

raffenet commented:
test:mpich/ch4/most
test:mpich/ch3/most

raffenet commented:
test:mpich/ch4/most
test:mpich/ch3/most

raffenet commented:
This is ready for a round of review. @sudheer9 is working on testing out the events functionality, which should hopefully turn into a few test cases we can add to this (or a follow-up) PR.

hzhou commented on configure.ac:
```diff
@@ -404,6 +404,12 @@ AC_ARG_ENABLE([mpit-pvars],
+AC_ARG_ENABLE([mpit-events],
+[ --enable-mpit-events=yes|no - Enable MPI_T events. (Default is "no"):
+        yes    - No events
+        no     - All available events
```
Did you mean the opposite?

raffenet replied:
Oops 🤦

raffenet replied:
Fixed.

hzhou commented on src/mpid/ch4/src/ch4r_recvq.c:
```diff
@@ -35,5 +37,16 @@ int MPIDIG_recvq_init(void)
+                          array_of_datatypes, array_of_displacements, 1,
+                          "message added to unexpected queue", MPI_T_BIND_NO_OBJECT, "CH4",
+                          &unexp_message_index);
+
```
Nice! I can use this! But to be useful, one would need to have dequeuing events as well or they wouldn't be able to monitor queue lengths -- the actual performance factor.

raffenet replied:
A companion dequeue event makes sense and could be easily added. Note that we do have PVARs (i.e. not events) that monitor queue lengths.

raffenet replied:
Added a dequeue event in latest push.

hzhou reviewed (commented):
some comments

raffenet commented:
test:mpich/ch4/most
test:mpich/ch3/most

hzhou commented on src/include/mpitimpl.h:
```diff
@@ -1238,5 +1241,48 @@ extern MPID_Thread_mutex_t mpi_t_mutex;
 extern void MPIR_T_env_finalize(void);
+extern void MPIR_T_events_finalize(void);
+
+typedef MPI_Count(MPIR_T_timestamp_fn) (void);
```
Is this the same as 
`typedef MPI_Count (*MPIR_T_timestamp_fn)(void);`
?`

raffenet replied:
Yes, can modify and make it clearer.

hzhou commented on src/include/mpitimpl.h:
```diff
@@ -1238,5 +1241,48 @@ extern MPID_Thread_mutex_t mpi_t_mutex;
+    MPI_Count max_ticks;
+
+    UT_hash_handle hh;          /* Makes this structure hashable */
+} MPIR_T_source_t;
```
A newline between the struct and function declarations?

hzhou replied:
Do we usually use `_t` suffix for `struct`s?  I usually associate `_t` types to some basic types such as `enums` or `integers`.

raffenet replied:
We do in the tools code apparently 😄. I wasn't aware of the convention beforehand. Not strongly tied to it.

hzhou commented on src/include/mpitimpl.h:
```diff
@@ -1238,5 +1241,48 @@ extern MPID_Thread_mutex_t mpi_t_mutex;
+    MPIR_T_bind_t bind;
+
+    UT_hash_handle hh;          /* Makes this structure hashable */
+} MPIR_T_event_t;
```
A newline separation here?

hzhou commented on src/mpi_t/events_impl.c:
```diff
@@ -178,25 +206,75 @@ int MPIR_T_event_get_timestamp_impl(MPI_T_event_instance event_instance,
+    if (event_registration->kind != MPIR_T_EVENT_REG) {
+        return MPI_T_ERR_INVALID_HANDLE;
+    }
+#endif
```
Should we consider to amend python script to add this validation to the MPI_T_event_registration  kind?

raffenet replied:
done

hzhou commented on src/mpi_t/events_impl.c:
```diff
@@ -194,13 +260,29 @@ int MPIR_T_event_get_num_impl(int *num_events)
+    if (event_instance->kind != MPIR_T_EVENT_INSTANCE) {
+        return MPI_T_ERR_INVALID_HANDLE;
+    }
+#endif
```
Do the validation in python script?

raffenet replied:
Will add.

raffenet replied:
done

hzhou commented on src/mpi_t/events_impl.c:
```diff
@@ -124,6 +129,7 @@ void MPIR_T_event_instance(int event_index, MPI_T_cb_safety cb_safety, void *dat
 
     return;
 }
+#endif /* HAVE_MPIT_EVENTS */
```
I don't quite understand why do we need to `#ifdef` out the entire function. Isn't `#ifdef` out the body sufficient?

raffenet replied:
After adding macro wrapper, changed the ifdef to just do the body of the function here.

hzhou commented on src/mpid/ch4/src/ch4r_recvq.h:
```diff
@@ -108,6 +112,7 @@ MPL_STATIC_INLINE_PREFIX MPIR_Request *MPIDIG_dequeue_unexp(int rank, int tag,
         req = curr->request;
         if (MPIDIG_match_unexp(rank, tag, context_id, req)) {
             DL_DELETE(*list, curr);
+            MPIR_T_event_instance(unexp_message_indices[1], MPI_T_CB_REQUIRE_MPI_RESTRICTED, &tag);
```
I see. You want to avoid the function call overhead altogether. Only `MPIR_T_even_instance` is relevant to the performance concern, right? I would rather use another macro wrapper, that defined to empty when events are disabled, than the current mechanism, which seems more mysterious to me.

raffenet replied:
So something like?
```
MPIR_T_event_instance (...) \
    do {} while (0)
```

hzhou replied:
I am more thinking about using a separate macro, e.g.
```
#ifdef HAVE_MPIT_EVENTS
#define MPIR_T_DO_EVENT MPIR_T_event_instance(...)
#else
#define MPIR_T_DO_EVENT do{}while(0)
#endif
```

`MPIR_T_event_instance` is a regular function, so reader may miss that it potentially is defined away. With a different macro name, the difference is explicit, and the reader can quickly check and understand the logics behind it.

raffenet replied:
Ah, yes. That's more like what the PVAR code does. I can make that change. `(...)` is from C99 onwards: https://en.cppreference.com/w/c/preprocessor/replace

raffenet replied:
Added MPIR_T_DO_EVENT.

hzhou commented on src/include/mpitimpl.h:
```diff
@@ -1316,6 +1316,10 @@ typedef struct MPIR_T_event_instance_s {
 void MPIR_T_event_instance(int event_index, MPI_T_cb_safety cb_safety, void *data);
+#else
+#define MPIR_T_event_instance(...)
+#endif
```
This is a bit mysterious. How does it work?

raffenet replied:
So right now you can either enable or disable events globally. When disabled (default), the event instance function calls become empty macros, as they are in performance critical paths.

hzhou replied:
I see. I was not familiar with the `...` syntax.

hzhou reviewed (commented):
Only minor comments.

hzhou commented on src/include/mpitimpl.h:
```diff
@@ -1238,5 +1243,83 @@ extern MPID_Thread_mutex_t mpi_t_mutex;
+    int num_elements;
+    MPIR_T_enum_t *enumtype;
+    char *desc;
+    MPIR_T_bind_t bind;
```
I may have missed it, where is `MPIR_T_bind_t` defined?

raffenet replied:
It's in `mpi.h` since it contains the definitions of standard user-visible enum values:
```
typedef enum MPIR_T_bind_t {
    /* don't name-shift this if/when MPI_T_ is accepted, this is an MPICH-only
     * extension */
    MPIX_T_BIND_INVALID = 0,

    /* arbitrarily shift values to aid debugging and reduce accidental errors */
    MPI_T_BIND_NO_OBJECT = 9700,
    MPI_T_BIND_MPI_COMM,
    MPI_T_BIND_MPI_DATATYPE,
    MPI_T_BIND_MPI_ERRHANDLER,
    MPI_T_BIND_MPI_FILE,
    MPI_T_BIND_MPI_GROUP,
    MPI_T_BIND_MPI_OP,
    MPI_T_BIND_MPI_REQUEST,
    MPI_T_BIND_MPI_WIN,
    MPI_T_BIND_MPI_MESSAGE,
    MPI_T_BIND_MPI_INFO
} MPIR_T_bind_t;
```

raffenet replied:
Just to be clear, this was a preexisting MPI_T definition.

raffenet commented:
test:mpich/ch4/most
test:mpich/ch3/most

raffenet commented:
I think I've addressed all comments. I added the handle validation changes as a separate commit since rebasing would be a bit time-consuming, but not impossible.

raffenet commented:
I don't yet have any tests in this PR, but can hack up a basic one along the lines of `test/mpi/mpi_t/mpit_vars.c`.

hzhou commented:
> 
> 
> I don't yet have any tests in this PR, but can hack up a basic one along the lines of `test/mpi/mpi_t/mpit_vars.c`.

That'll be nice. Yeah, a basic one, for now, is good enough.

raffenet commented:
test:mpich/ch4/most
test:mpich/ch3/most

hzhou commented on src/mpi_t/events_impl.c:
```diff
@@ -121,6 +126,7 @@ void MPIR_T_event_instance(int event_index, MPI_T_cb_safety cb_safety, void *dat
             event_reg->dropped_count++;
         }
     }
+#endif /* HAVE_MPIT_EVENTS */
```
Now that `MPIR_T_DO_EVENT` takes care of the turning off events, is it still necessary to comment out the body of these functions?

raffenet replied:
Event_instance doesn't need to be. For the internal registration functions, I suppose I can add macro wrappers for them, too. That way the function can exist normally, but no events or sources will be reported when disabled.

hzhou replied:
I was thinking that leaving the registration active would be harmless, but I guess that may confuse the user. I guess we can just leave it as is for now.

raffenet replied:
If we report events and sources, the user may be confused why the callbacks never execute. On the PVAR side, we report no PVARs available unless explicitly enabled.

raffenet replied:
Added wrappers for all internal APIs and removed `#ifdefs` from within the functions.

raffenet replied:
Hmm, so turning the registration functions into empty macros causes warnings for unused variables:
```
In directory: /var/lib/jenkins-slave/workspace/mpich-warnings/compiler/gcc-4/config/ch4-ucx/label/centos64_review
  CC       src/mpid/ch4/netmod/ucx/lib_libmpi_la-ucx_init.lo
src/mpid/ch4/src/ch4r_recvq.c: In function ‘MPIDIG_recvq_init’:
src/mpid/ch4/src/ch4r_recvq.c:45:14: warning: unused variable ‘array_of_displacements’ [-Wunused-variable]
     MPI_Aint array_of_displacements[1] = { 0 };
              ^
src/mpid/ch4/src/ch4r_recvq.c:44:18: warning: unused variable ‘array_of_datatypes’ [-Wunused-variable]
     MPI_Datatype array_of_datatypes[1] = { MPI_INT };
```
I'll revert those wrappers for now.

hzhou reviewed (approved):
LGTM.

raffenet commented:
test:mpich/ch4/most
test:mpich/ch3/most

raffenet commented:
test:mpich/ch4/most
test:mpich/ch3/most
